### PR TITLE
Convert ExecuteNonQueryAsync to use async context object

### PR DESCRIPTION
**Pull request by Wraith2**
fixes https://github.com/dotnet/SqlClient/issues/1682

This addresses the incorrect handling of cancellation token registrations in `SqlCommand` methods `ExecuteNonQuery` and `ExecuteReader`, I didn't change `ExecuteScalar` because it should be done in a separate PR and it's just `ExecuteReader` wearing a fake nose a glasses, people should use `ExexuteReader`

Background information: There are a number of async api calls on `SqlDataReader` which have been changed to avoid the use of language provided lambda closures and instead use explicit context objects. This process makes the code more complicated because it splits apart the code into multiple functions so that instance delegate allocations can be avoided. This pattern is no longer needed if we use language provided static lambdas and the code can be much cleaner and better contained. This PR converts the ExecuteNonQuery function to use this explicity context pattern and uses static lambdas to doi t.

This PR also changes all existing context objects to use an explicit type for the IDispoable so that we can avoid boxing the disposable. This change forces any existing context types which were declared like this: `AAsyncContext<TOwner,TResult>` to be split apart into two types, 
1) `AAsyncContext<TOwner,TResult,TDisposable>` which has a strong disposable type and is used at call sites that know what the disposable is, mostly setup locations.  
2)  `AAsyncBaseContext<TOwner,TResult>` which is used in the async functions that don't need to know what the type of the disposable is so we can avoid duplicating functions like `InvokeAsync`

The specific changes to fix the issue reported are 
1) change `InternalExecuteNonQueryAsync` to use an explicity context object 
2) make disposable idempotent by including a disposed flag. Previously it was not possible to reach disposable from more than one path so knowledge of disposal was implicit, now it is explicit.
3) call context dispose from catch blocks ensuring that the registration is correctly cleaned up on all paths

/cc @olegkv

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/AAsyncCallContext.cs [hunk 1]
```diff
@@ -17,38 +17,69 @@ namespace Microsoft.Data.SqlClient
...
+        {
+            base.Set(owner, source);
+            _disposable = disposable;
+        }
+
+        protected override void DisposeCore()
+        {
+            TDisposable copyDisposable = _disposable;
+            _disposable = default;
+            _isDisposed = true;
```
Comment (olegkv) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/AAsyncCallContext.cs:45]: isDisposed is already set by parent class in ClearCore()
ReadAsyncCallContext class is another descendant which seems to do it in more consistent way, it does not touch _isDisposed and relies on parent class instead. Maybe isDisposed should be made private..

**Comment by lcheunglci** [comment]
@Wraith2 Looks like there's a merge conflict. Let us know if you would like us to resolve it for you?

**Comment by David-Engel** [comment]
> Looks like there's a merge conflict. Let us know if you would like us to resolve it for you.

I took care of it. It was a small change from #1781.

**Comment by David-Engel** [comment]
That'll teach me to use the GitHub conflict UI. I broke it. I'm looking at it.

**Comment by Wraith2** [comment]
Rebased, checked it still makes sense and addressed the feedback above.

**Review by David-Engel — APPROVED** [review]

**Review by JRahnama — APPROVED** [review]

**Pull request merged**